CIP-optimized-uplc-serialization/README.md

```diff
@@ -0,0 +1,446 @@
+
+two options are provided as alternative to the current design;
+
+option 1 helps backwards compatibitlity since it doesn't changes the serialization of same tags; but takes 4 bits per tag when 3 bits are enough to express all the alternatives. 
+
+this is why option 2 is present; allowing the 3 bits per tag serialization at the cost of missing bits for expansion in the evenience new constant types will be added; this cost needs to be discussed with the comunity since 1 bit per tag is a considerable improvment for complex types and is a 20% improvement on single element types. 
+
+### ```ByteString```s encoded as unsinged integer followed by the bytes
+
+ByteStrings are probably the values that have the most important impact on the serialized output;
+
+ByteStrings are used to represent a wide variety of types (```TokenName```, ```ValidatorHash```, ```PubKeyHash```, ```CurrencySymbol```, etc. are all ```ByteString``` wrappers ) ence widely used in smart contracts.
+
+citing the plutus core specificaiton draft released in june 2022 (latest specification at the time of writing)
+
+> Bytestrings are encoded by dividing them into nonempty blocks of up to 255 bytes and emitting each block in sequence.
+>
+> Each block is preceded by a single unsigned byte containing its length, and the end of the encoding is marked by a zero-length block (so the empty bytestring is encoded just as a zero-length block).
+>
+> Before emitting a bytestring, the preceding output is padded so that its length (in bits) is a multiple of 8;
+>
+> if this is already the case a single padding byte is still added; this ensures that contents of the bytestring are aligned to byte boundaries in the output.
+
+given the importance of bytestrings too much hoverhead is introduced; currently a ```ByteString``` serializes as
+
+```
+1) some padding that, depending on the context, may take up to 1 byte
+2) 1 (byte * chunk) indicating how many bytes are in the very next chunk
+3) 0 to 255 meaningful (bytes * chunk)
+4) 1 useless byte indicating no more chiunks are following
+```
+this implies ```(~1) + #chunks + 1``` meaningless bytes are added per ```ByteString```
+
+in the descripton above
+- step ```1``` allows for an easy serialization and deserialization but doesn't carries any meaningful informations; given the importance of ByteStrings it should be removed at the cost of an added shift while serializing/deserializing the value
+- step ```4``` can be removed as it is simply useless
+- step ```3``` can be optimized using unsigned integer encoding
+
+the new algorithm for ```ByteString``` serialization would consist of two parts:
+```
+1) Unsigned (indefinite length) Integer indicating how many bytes ( or chunks of 8 bits ) are following
+2) as many bytes as specified in the Unsigned Integer at step 1
+```
+
+using the new algorithm the ```ByteString``` serialization space complexity goes form ```O(n)``` to ```O(log n)``` where ```n``` is the number of bytes in the ```ByteString``` 
```

Comment:
Hmmm.  I need to think more carefully about this.  It's a while since I've thought about bytestring encodings and I'll need to remind myself of the details of how we currently do it.